staging/src/k8s.io/dynamic-resource-allocation/structured/namedresources/cel/semverlib.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"github.com/blang/semver/v4"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// SemVer provides a CEL function library extension for [semver.Version].
+//
+// semver
+//
+// Converts a string to a SemVer or results in an error if the string is not a valid SemVer. Refer
+// to semver.SemVer documentation for information on accepted patterns.
+//
+//	semver(<string>) <SemVer>
+//
+// Examples:
+//
+//	semver('1.0.0') // returns a SemVer
+//	semver('0.1.0-alpha.1') // returns a SemVer
+//	semver('200K') // error
+//	semver('Three') // error
+//	semver('Mi') // error
+//
+// isSemVer
+//
+// Returns true if a string is a valid SemVer. isSemVer returns true if and
+// only if semver does not result in error.
+//
+//	isSemVer( <string>) <bool>
+//
+// Examples:
+//
+//	isSemVer('1.0.0') // returns true
+//	isSemVer('v1.0') // returns true (tolerant parsing)
+//	isSemVer('hello') // returns false
+//
+// Conversion to Scalars:
+//
+//   - major/minor/patch: return the major version number as int64.
+//
+//     <SemVer>.major() <int>
+//
+// Examples:
+//
+// semver("1.2.3").major() // returns 1
+//
+// Comparisons
+//
+//   - isGreaterThan: Returns true if and only if the receiver is greater than the operand
+//
+//   - isLessThan: Returns true if and only if the receiver is less than the operand
+//
+//   - compareTo: Compares receiver to operand and returns 0 if they are equal, 1 if the receiver is greater, or -1 if the receiver is less than the operand
+//
+//
+//     <SemVer>.isLessThan(<semver>) <bool>
+//     <SemVer>.isGreaterThan(<semver>) <bool>
+//     <SemVer>.compareTo(<semver>) <int>
+//
+// Examples:
+//
+// semver("1.2.3").compareTo(semver("1.2.3")) // returns 0
+// semver("1.2.3").compareTo(semver("2.0.0")) // returns -1
+// semver("1.2.3").compareTo(semver("0.1.2")) // returns 1
```

Comment:
cc @cici37 

Library looks really good to me. 

Sanity check- do all the functions appear to have O(1) running time? A large string input into `semver`  is probably the theoretically worst case, but in practice, can we consider that O(1)? If so we don't need to do any cost system tuning.  I think I'm okay with that. Does anyone object?

Reply:
Over in https://github.com/kubernetes/kubernetes/pull/123664 (my attempt to get this into the base library) I used the same costs as for the quantity type. They are very comparable (short string input, simple comparisons).

I think in the context of this alpha API we don't need to worry about cost estimates yet. We don't know how complex real-world CEL expressions will be.

Reply:
> A large string input into semver is probably the theoretically worst case

`string → semver` and `issemver(string)` would be the expensive operations, right? the longer the string, the more expensive?

once we have semvers, the comparisons seem *very* cheap

Reply:
Adding `semver` and `isSemver` to https://github.com/kubernetes/kubernetes/blob/46f017a90b54b5abfc0a66a6ecf19e304aa7da95/staging/src/k8s.io/apiserver/pkg/cel/library/cost.go#L54 is all we need t o do. See `url` and `isUrl` for reference.

Reply:
Yes, that `cost := uint64(math.Ceil(float64(actualSize(args[0])) * common.StringTraversalCostFactor))` is what I did for `semver` and `isSemver`, the same as for `quantity` and the others which take a string as input.

Reply:
Perfect. Thx!